haystack_experimental/components/agents/agent.py

```diff
@@ -0,0 +1,264 @@
+        self.pipeline.add_component(
+            instance=context_joiner, name="context_joiner"
+        )
+
+        # Connect components
+        self.pipeline.connect("joiner.value", "generator.messages")
+        self.pipeline.connect("generator.replies", "router.llm_messages")
+        self.pipeline.connect("joiner.value", "router.original_messages")
+        self.pipeline.connect("generator.replies", "tool_invoker.messages")
+        self.pipeline.connect("tool_invoker.tool_messages", "router.tool_messages")
+        self.pipeline.connect("router.continue", "joiner.value")
+        self.pipeline.connect(
+            "tool_invoker.state", "context_joiner.value"
+        )
+        self.pipeline.connect(
+            "context_joiner.value", "tool_invoker.state"
+        )
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Serialize the component to a dictionary.
+
+        :return: Dictionary with serialized data
+        """
+        return default_to_dict(
+            self,
+            model=self.model,
+            generation_kwargs=self.generation_kwargs,
+            tools=[t.to_dict() for t in self.tools],
+            api_key=self.api_key.to_dict(),
+            system_prompt=self.system_prompt,
+            exit_condition=self.exit_condition,
+            state_schema=_schema_to_dict(self.state_schema),
+        )
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "Agent":
+        """
+        Deserialize the component from a dictionary.
+
+        :param data: Dictionary to deserialize from
+        :return: Deserialized component
+        """
+        init_params = data.get("init_parameters", {})
+
+        # Deserialize type annotations
+        if "state_schema" in init_params:
+            init_params["state_schema"] = _schema_from_dict(init_params["state_schema"])
+
+        if "tools" in init_params:
+            init_params["tools"] = [Tool.from_dict(t) for t in init_params["tools"]]
+
+        deserialize_secrets_inplace(init_params, ["api_key"])
+
+        return default_from_dict(cls, data)
+
+    def run(self, messages: List[ChatMessage], **kwargs) -> Dict[str, Any]:
+        """
+        Process messages and execute tools until the exit condition is met.
+
+        :param messages: List of chat messages to process
+        :param kwargs: Additional keyword arguments matching the defined input types
+        :return: Dictionary containing messages and outputs matching the defined output types
+        """
+        state = State(schema=self.state_schema, data=kwargs)
+
+        if self.system_prompt is not None:
+            messages = [ChatMessage.from_system(self.system_prompt)] + messages
+
+        self.pipeline.warm_up()
+
+        result = self.pipeline.run(
+            data={
+                "joiner": {"value": messages},
+                "context_joiner": {"value": state},
+            },
+            include_outputs_from={"context_joiner"},
+        )
```

Comment:
@mathislucka just a general question. Let's say I wanted to see the whole "thinking" process of the Agent, so all intermediate outputs before the exit condition is met. Is there an easy way to get this (e.g. creating the right kind of State schema?)

Reply:
The agent returns all chat messages including any intermediate tool calls, tool results, etc.

I also thought about adding the messages to state by default but came to the conclusion that it isn't needed because we already get them as a proper output.

It might be relevant to add them to State when we think about interrupting an agent (breakpoints) and later picking back up from where it left off. I'd say we keep it like it is right now and can still change it later on.